contracts/margin/external/AuctionProxy.sol

```diff
@@ -0,0 +1,176 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+pragma solidity 0.4.24;
+pragma experimental "v0.5.0";
+
+import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import { Margin } from "../Margin.sol";
+import { ZeroExExchangeInterface } from "../../external/0x/ZeroExExchangeInterface.sol";
+import { MathHelpers } from "../../lib/MathHelpers.sol";
+import { TokenInteract } from "../../lib/TokenInteract.sol";
+import { ZeroExV1Parser } from "../../lib/ZeroExV1Parser.sol";
+
+
+/**
+ * @title AuctionProxy
+ * @author dYdX
+ *
+ * Contract for allowing anyone to close a Dutch Auction
+ */
+contract AuctionProxy is
+    ZeroExV1Parser
+{
+    using TokenInteract for address;
+    using SafeMath for uint256;
+
+    // ============ Structs ============
+
+    struct Position {
+        address heldToken;
+        address owedToken;
+        address tokenContract;
+        uint256 principal;
+        uint256 owedTokenOwed;
+        uint256 collateralAmount;
+    }
+
+    // ============ State Variables ============
+
+    address public DYDX_MARGIN;
+    address public ZERO_EX_V1_EXCHANGE;
+
+    // ============ Constructor ============
+
+    constructor(
+        address margin,
+        address zeroExExchange
+    )
+        public
+    {
+        DYDX_MARGIN = margin;
+        ZERO_EX_V1_EXCHANGE = zeroExExchange;
+    }
+
+    // ============ Public Functions ============
+
+    function closePosition(
+        bytes32 positionId,
+        address dutchAuction,
+        address exchangeWrapper,
+        bytes   orderData
+    )
+        external
+        returns (uint256)
+    {
+        Margin margin = Margin(DYDX_MARGIN);
+        Position memory position = parsePosition(margin, positionId);
+
+        uint256 maxCloseAmount = getMaxCloseAmount(position, orderData);
+
+        margin.closePosition(
+            positionId,
+            maxCloseAmount,
+            dutchAuction,
+            exchangeWrapper,
+            true, // payoutInHeldToken
+            orderData
+        );
+
+        // give all tokens to the token contract
+        uint256 heldTokenAmount = position.heldToken.balanceOf(address(this));
+        position.heldToken.transfer(position.tokenContract, heldTokenAmount);
+
+        return maxCloseAmount;
+    }
+
+    // ============ Private Functions ============
+
+    function parsePosition (
+        Margin margin,
+        bytes32 positionId
+    )
+        internal
+        view
+        returns (Position memory)
+    {
+        Position memory position;
+        position.heldToken = margin.getPositionHeldToken(positionId);
```

Comment:
why make all these calls individually? I think it'll be simpler to just use `MarginHelper#getPosition`. I'm not sure which is more gas efficient (more storage loads versus more contract calls)